src/sources.cr

```diff
@@ -76,7 +103,68 @@ class LiveSource < Source
   def set_last_message(@last_message)
   end
 
-  def last_data : Hash(String, JSON::Any)
-    @last_message.as_h
+  def last_data : SourceData
+    @current_data
+  end
+end
+
+class ChroniclerSource < Source
+  property tx : Channel({String, SourceData})
+  property ident : String
+  property clients : Int32 = 0
+  property current_data : SourceData
+  property cached_messages : Array({Time, SourceData}) = Array({Time, SourceData}).new
+  property last_time : Time
+  property running : Bool = true
+
+  def initialize(@last_time, @ident : String, @tx : Channel({String, SourceData}))
+    @current_data = SourceData.new
+    spawn do
+      while @running
+        if cached_messages.size == 0
+          fetch_messages
+        end
+
+        break if cached_messages.size == 0
+
+        next_message = @cached_messages.delete_at 0
+
+        sleep next_message[0] - @last_time
+        @last_time = next_message[0]
+        @tx.send({ident, next_message[1]})
+      end
+    end
+  end
+
+  def add_client
+    @clients += 1
+  end
+
+  def rm_client
+    @clients -= 1
+  end
+
+  def n_clients
+    @clients
+  end
+
+  def close
+    @running = false
+  end
+
+  def fetch_messages
+    params = URI::Params.encode({"type" => "Stream", "count" => "30", "order" => "asc", "after" => @last_time.to_rfc3339})
+    response = HTTP::Client.get URI.new(scheme: "https", host: "api.sibr.dev", path: "/chronicler/v2/versions", query: params)
```

Comment:
It would be nice if this didn't hardcode the chron address, so that it could be used with VCR